File: lib/operation-stack.ts
import * as autoscaling from "aws-cdk-lib/aws-autoscaling";
import * as autoscaling_hooktargets from "aws-cdk-lib/aws-autoscaling-hooktargets";
import * as backup from "aws-cdk-lib/aws-backup";
import * as cdk from "aws-cdk-lib";
import * as cw_actions from "aws-cdk-lib/aws-cloudwatch-actions";
import * as devopsguru from "aws-cdk-lib/aws-devopsguru";
import * as lambda from "aws-cdk-lib/aws-lambda";
import * as sns from "aws-cdk-lib/aws-sns";
import * as subscriptions from "aws-cdk-lib/aws-sns-subscriptions";

import { Construct } from "constructs";
import { Environment } from "./environment";

import { ComputationStack } from "./computation-stack";
import { DomainStack } from "./domain-stack";
import { EmailStack } from "./email-stack";
import { PersistentStorageStack } from "./persistent-storage-stack";
import { ServerStack } from "./server-stack";

type OperationStackProps = cdk.StackProps & {
  environment: Environment;
  domainStack: DomainStack;
  computationStack: ComputationStack;
  emailStack: EmailStack;
  persistentStorageStack: PersistentStorageStack;
  serverStack: ServerStack;
};

export class OperationStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props: OperationStackProps) {
    super(scope, id, props);

    const environment = props.environment;

    // Backup

    let backupPlan = backup.BackupPlan.dailyWeeklyMonthly5YearRetention(
      this,
      `${this.stackName}-BackupPlan`,
      props.persistentStorageStack.backupVault
    );

    backupPlan.addSelection("EmailSuppressionListBackup", {
      resources: [
        backup.BackupResource.fromDynamoDbTable(
          props.persistentStorageStack.emailSuppressionList
        ),
      ],
    });

    backupPlan.addSelection("ServerBlockStorageBackup", {
      resources: [
        backup.BackupResource.fromArn(
          `arn:${this.partition}:ec2:${this.region}:${this.account}:volume/${props.serverStack.containerInstance.volumeId}`
        ),
      ],
    });

    // Alarms

    let systemOperationsTopic = new sns.Topic(this, "SystemOperationsTopic");

    systemOperationsTopic.addSubscription(
      new subscriptions.EmailSubscription(
        `admins@${props.domainStack.domainName}`
      )
    );

    let allConcurrentExecutionsMetric =
      lambda.Function.metricAllConcurrentExecutions({
        period: cdk.Duration.minutes(5),
        statistic: "max",
      });

    let allConcurrentExecutionsAlarm =
      allConcurrentExecutionsMetric.createAlarm(
        this,
        "AllConcurrentExecutionsAlarm",
        { evaluationPeriods: 1, threshold: 100 }
      );

    allConcurrentExecutionsAlarm.addAlarmAction(
      new cw_actions.SnsAction(systemOperationsTopic)
    );

    let serverAutoScalingGroup =
      props.serverStack.service.cluster.autoscalingGroup;

    if (!serverAutoScalingGroup)
      throw Error("Server autoscaling group is not defined");

    new autoscaling.LifecycleHook(this, "InstanceTerminationLifecycleHook", {
      autoScalingGroup: serverAutoScalingGroup,
      lifecycleTransition: autoscaling.LifecycleTransition.INSTANCE_TERMINATING,
      notificationTarget: new autoscaling_hooktargets.TopicHook(
        systemOperationsTopic
      ),
    });

    // Automatic monitoring

    new devopsguru.CfnNotificationChannel(
      this,
      "OperationsNotificationChannel",
      { config: { sns: { topicArn: systemOperationsTopic.topicArn } } }
    );

    let stacks: cdk.Stack[] = [
      props.domainStack,
      props.computationStack,
      props.emailStack,
      props.persistentStorageStack,
      props.serverStack,
    ];

    new devopsguru.CfnResourceCollection(this, "ResourceCollection", {
      resourceCollectionFilter: {
        cloudFormation: { stackNames: stacks.map((stack) => stack.stackName) },
      },
    });
  }
}

File: nitecap/operation_stack.py
import tldextract

import aws_cdk as cdk
import aws_cdk.aws_autoscaling as autoscaling
import aws_cdk.aws_autoscaling_hooktargets as autoscaling_hooktargets
import aws_cdk.aws_backup as backup
import aws_cdk.aws_cloudwatch as cloudwatch
import aws_cdk.aws_cloudwatch_actions as cw_actions
import aws_cdk.aws_devopsguru as devopsguru
import aws_cdk.aws_lambda as λ
import aws_cdk.aws_sns as sns
import aws_cdk.aws_sns_subscriptions as subscriptions

from constructs import Construct

from .configuration import Configuration
from .computation_stack import ComputationStack
from .domain_stack import DomainStack
from .email_stack import EmailStack
from .persistent_storage_stack import PersistentStorageStack
from .server_stack import ServerStack


class OperationStack(cdk.Stack):
    def __init__(
        self,
        scope: Construct,
        construct_id: str,
        *,
        configuration: Configuration,
        domain_stack: DomainStack,
        computation_stack: ComputationStack,
        email_stack: EmailStack,
        persistent_storage_stack: PersistentStorageStack,
        server_stack: ServerStack,
        **kwargs,
    ) -> None:
        super().__init__(scope, construct_id, **kwargs)

        # Backup

        backup_plan = backup.BackupPlan.daily_monthly1_year_retention(
            self, f"{self.stack_name}-BackupPlan", persistent_storage_stack.backup_vault
        )

        backup_plan.add_selection(
            "EmailSuppressionListBackup",
            resources=[
                backup.BackupResource.from_dynamo_db_table(
                    persistent_storage_stack.email_suppression_list
                )
            ],
        )

        backup_plan.add_selection(
            "ServerBlockStorageBackup",
            resources=[
                backup.BackupResource.from_arn(
                    f"arn:{self.partition}:ec2:{self.region}:{self.account}:volume/{server_stack.container_instance.volume_id}"
                )
            ],
        )

        # Alarms

        system_operations_topic = sns.Topic(self, "SystemOperationsTopic")

        subdomain, domain, suffix = tldextract.extract(configuration.domain_name)

        system_operations_topic.add_subscription(
            subscriptions.EmailSubscription(f"admins@{domain}.{suffix}")
        )

        all_concurrent_executions_metric = λ.Function.metric_all_concurrent_executions(
            period=cdk.Duration.minutes(5), statistic="max"
        )

        all_concurrent_executions_alarm = all_concurrent_executions_metric.create_alarm(
            self,
            "AllConcurrentExecutionsAlarm",
            evaluation_periods=1,
            threshold=100,
            treat_missing_data=cloudwatch.TreatMissingData.IGNORE,
        )

        all_concurrent_executions_alarm.add_alarm_action(
            cw_actions.SnsAction(system_operations_topic)
        )

        autoscaling.LifecycleHook(
            self,
            "InstanceTerminationLifecycleHook",
            auto_scaling_group=server_stack.service.cluster.autoscaling_group,
            lifecycle_transition=autoscaling.LifecycleTransition.INSTANCE_TERMINATING,
            notification_target=autoscaling_hooktargets.TopicHook(
                system_operations_topic
            ),
        )

        # Automatic monitoring

        devopsguru.CfnNotificationChannel(
            self,
            "OperationsNotificationChannel",
            config=devopsguru.CfnNotificationChannel.NotificationChannelConfigProperty(
                sns=devopsguru.CfnNotificationChannel.SnsChannelConfigProperty(
                    topic_arn=system_operations_topic.topic_arn
                )
            ),
        )

        devopsguru.CfnResourceCollection(
            self,
            "ResourceCollection",
            resource_collection_filter=devopsguru.CfnResourceCollection.ResourceCollectionFilterProperty(
                cloud_formation=devopsguru.CfnResourceCollection.CloudFormationCollectionFilterProperty(
                    stack_names=[
                        domain_stack.stack_name,
                        persistent_storage_stack.stack_name,
                        email_stack.stack_name,
                        computation_stack.stack_name,
                        server_stack.stack_name,
                    ]
                )
            ),
        )
